src/lib.rs

```diff
@@ -368,36 +367,81 @@ pub struct Socket {}
+    },
+    /// Failed to parse hostname. The invalid hostname that caused this error is found in
+    /// `.host`. An optional explanation may be found in `.detail`. The lower level
+    /// `url::ParseError` is contained in `.parse_error`.
+    InvalidHost {
+        host: String,
+        detail: Option<String>,
+        parse_error: url::ParseError,
+    },
+    /// Invalid option or value specified, may only be invalid in context. The invalid option
+    /// is found in `.option`, while the invalid value is found in `.value`. An optional
+    /// explanation providing context may be found in `.detail`.
+    InvalidOption {
+        option: String,
+        value: String,
+        detail: Option<String>,
+    },
+    /// Invalid wait time specified. The minimum wait time and maximum wait time are found in
+    /// `.min_wait` and `.max_wait` respectively. An optional explanation providing context may
+    /// be found in `.detail`.
+    InvalidWaitTime {
+        min_wait: usize,
+        max_wait: usize,
+        detail: Option<String>,
+    },
+    /// Invalid weight specified. The invalid weight value is found in `.weight`. An optional
+    // explanation providing context may be found in `.detail`.
+    InvalidWeight {
+        weight: usize,
+        detail: Option<String>,
+    },
+    /// `GooseAttack` has no `GooseTaskSet` defined. An optional explanation may be found in
+    /// `.detail`.
+    NoTaskSets {
+        detail: Option<String>,
+    },
 }
 
+// Define how to display errors.
 impl fmt::Display for GooseError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(&self, f)
+    }
+}
+
+// Define the lower level source of this error, if any.
+impl std::error::Error for GooseError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
         match *self {
-            GooseError::Io(ref err) => err.fmt(f),
-            GooseError::Reqwest(ref err) => err.fmt(f),
-            GooseError::FeatureNotEnabled => write!(f, "Compile time feature not enabled."),
-            GooseError::InvalidHost => write!(f, "Unable to parse host."),
-            GooseError::InvalidOption => write!(f, "Invalid option specified."),
-            GooseError::InvalidWaitTime => write!(f, "Invalid wait time specified."),
-            GooseError::InvalidWeight => write!(f, "Invalid weight specified."),
-            GooseError::NoTaskSets => write!(f, "No task sets defined."),
+            GooseError::Io(ref source) => Some(source),
+            GooseError::Reqwest(ref source) => Some(source),
+            GooseError::FeatureNotEnabled { .. } => None,
+            GooseError::InvalidHost {
+                ref parse_error, ..
+            } => Some(parse_error),
+            GooseError::InvalidOption { .. } => None,
+            GooseError::InvalidWaitTime { .. } => None,
+            GooseError::InvalidWeight { .. } => None,
+            GooseError::NoTaskSets { .. } => None,
```

Comment:
I think you can also match `_ => None` if you don't want to enumerate all ...

Reply:
Excellent point, easier to maintain as we add more errors.